packages/jest-haste-map/src/crawlers/watchman.js

```diff
@@ -18,126 +18,121 @@ import H from '../constants';
-        client.command(args, (error, result) => {
-          if (error) {
-            reject(error);
-          } else {
-            resolve(result);
-          }
-        });
-      });
+  const clocks = data.clocks;
+  let files = data.files;
 
-    const clocks = data.clocks;
-    let files = data.files;
+  try {
+    const watchmanRoots = new Map();
+    for (const root of roots) {
+      const response = await cmd('watch-project', root);
+      const existing = watchmanRoots.get(response.watch);
+      // A root can only be filtered if it was never seen with a relative_path before
+      const canBeFiltered = !existing || existing.length > 0;
 
-    return Promise.all(roots.map(root => cmd('watch-project', root)))
-      .then(responses =>
-        Promise.all(
-          Array.from(new Set(responses.map(response => response.watch))).map(
-            root => {
-              // Build an expression to filter the output by the relevant roots.
-              const dirExpr = (['anyof']: Array<string | Array<string>>);
-              posixRoots.forEach(subRoot => {
-                if (isDescendant(root, subRoot)) {
-                  dirExpr.push(['dirname', path.relative(root, subRoot)]);
-                }
-              });
-              const expression = [
-                'allof',
-                ['type', 'f'],
-                ['anyof'].concat(
-                  extensions.map(extension => ['suffix', extension]),
-                ),
-              ];
-              if (dirExpr.length > 1) {
-                expression.push(dirExpr);
-              }
-              const fields = ['name', 'exists', 'mtime_ms'];
+      if (canBeFiltered) {
+        if (response.relative_path) {
+          watchmanRoots.set(
+            response.watch,
+            (existing || []).concat(response.relative_path),
+          );
+        } else {
+          // Make the filter directories an empty array to signal that this root
+          // was already seen and needs to be watched for all files/directories
+          watchmanRoots.set(response.watch, []);
+        }
+      }
+    }
 
-              const query = clocks[root]
-                ? // Use the `since` generator if we have a clock available
-                  {expression, fields, since: clocks[root]}
-                : // Otherwise use the `suffix` generator
-                  {expression, fields, suffix: extensions};
-              return cmd('query', root, query).then(response => ({
-                response,
-                root,
-              }));
-            },
-          ),
-        ).then(pairs => {
-          // Reset the file map if watchman was restarted and sends us a list of
-          // files.
-          if (pairs.some(pair => pair.response.is_fresh_instance)) {
-            files = Object.create(null);
-          }
+    let shouldReset = false;
+    const watchmanFileResults = new Map();
+    for (const [root, directoryFilters] of watchmanRoots) {
```

Comment:
not parallelizing means that you have to wait for all watch-project commands before you can send the queries.